xapian-applications/omega/handler_gmime.cc

```diff
@@ -0,0 +1,183 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+#include <config.h>
+#include "handler.h"
+
+#include "myhtmlparse.h"
+#include <string.h>
+
+#include <glib.h>
+#include <gmime/gmime.h>
+
+using namespace std;
+
+const unsigned SIZE = 8192;
+
+static void
+extract_html(const string& text, string& charset, string& dump)
+{
+    MyHtmlParser parser;
+    if (charset.empty())
+	charset = "UTF-8";
+    try {
+	parser.ignore_metarobots();
+	parser.parse_html(text, charset, false);
+	dump += parser.dump;
+    } catch (const string& newcharset) {
+	parser.reset();
+	parser.ignore_metarobots();
+	parser.parse_html(text, newcharset, true);
+	dump += parser.dump;
+    } catch (...) {
+	return;
+    }
+}
+
+size_t decode(unsigned char* text, size_t len, GMimeContentEncoding encoding)
+{
+    unsigned char buffer[SIZE];
+    int state;
+    guint32 save;
+    switch (encoding) {
+	case GMIME_CONTENT_ENCODING_BASE64: {
+	    len = g_mime_encoding_base64_decode_step(text, len, buffer,
+						     &state, &save);
+	    memcpy(text, buffer, len);
+	    break;
+	}
+	case GMIME_CONTENT_ENCODING_UUENCODE: {
+	    len = g_mime_encoding_uudecode_step(text, len, buffer,
+						&state, &save);
+	    memcpy(text, buffer, len);
+	    break;
+	}
+	case GMIME_CONTENT_ENCODING_QUOTEDPRINTABLE: {
+	    len = g_mime_encoding_quoted_decode_step(text, len, buffer,
+						     &state, &save);
+	    memcpy(text, buffer, len);
+	    break;
+	}
+	default:
+	    break;
+    }
+    return len;
+}
+
+bool
+parser_content(GMimeObject* me, string& dump)
+{
+    GMimeContentType* ct = g_mime_object_get_content_type(me);
+    if (GMIME_IS_MULTIPART(me)) {
+	GMimeMultipart* mpart = (GMimeMultipart*)me;
+	string subtype = g_mime_content_type_get_media_subtype(ct);
+	int count = g_mime_multipart_get_count(mpart);
+	if (subtype == "alternative") {
+	    for (int i = 0; i < count; ++i) {
+		GMimeObject* part = g_mime_multipart_get_part(mpart, i);
+		if (parser_content(part, dump))
+		    return true;
+	    }
+	} else {
+	    bool ret = false;
+	    for (int i = 0; i < count; ++i) {
+		GMimeObject* part = g_mime_multipart_get_part(mpart, i);
+		ret |= parser_content(part, dump);
+	    }
+	    return ret;
+	}
+    } else if (GMIME_IS_PART(me)) {
+	GMimePart* part = (GMimePart*)me;
+	GMimeDataWrapper* content = g_mime_part_get_content_object(part);
+	string type = g_mime_content_type_get_media_type(ct);
+	string subtype = g_mime_content_type_get_media_subtype(ct);
+	string charset(g_mime_object_get_headers(me));
+	size_t end, start;
+	start = charset.find("charset=");
```

Comment:
I was looking at this again as it seemed surprising that gmime didn't handle pulling the charset out from the header for us.  I think we ought to be using this function:

https://developer.gnome.org/gmime/stable/GMimeObject.html#g-mime-object-get-content-type-parameter

I tried this:

```
        string charset;  
        {                                                                            
            const char* p = g_mime_object_get_content_type_parameter(me,  
                                                                     "charset");  
            if (p) charset = g_mime_charset_canon_name(p);  
        }  
```

And it seems to give `UTF-8` in a simple test, same as your current code here does.